packages/teuchos/numerics/src/Teuchos_SerialBandDenseMatrix.hpp

```diff
@@ -1075,6 +1070,16 @@ void SerialBandDenseMatrix<OrdinalType, ScalarType>::copyMat(
   }
 }
 
+#ifndef TEUCHOS_HIDE_DEPRECATED_CODE
+/// \brief Print the given SerialBandDenseMatrix to the given output stream.
+template<typename OrdinalType, typename ScalarType>
+std::ostream& operator<< (std::ostream& os, const Teuchos::SerialBandDenseMatrix<OrdinalType, ScalarType>& obj)
```

Comment:
@rhoope, is function is not actually deprecated.  Technically speaking, to deprecate, you need to apply the macro `TEUCHOS_DEPRECATED` as described at:

* https://tribits.org/doc/TribitsDevelopersGuide.html#deprecating-c-c-classes-structs-functions-typdefs-etc

Otherwise, users will not know this function is deprecated and therefore will not know to stop using it.

Reply:
@bartlettroscoe I can add the macro to alert clients to the deprecation of the code.  What I'm unclear on is the idea of adding new headers; won't this break backward compatibility by requiring clients to include the new files?

Reply:
@rhoope, the problem is that if you add the `TEUCHOS_DEPRECATED` macro then you need to provide a way for existing users to remove the usage of the deprecated code.  How can they do this?  We don't expect them to have to change to raw `obj.print()` do we?  That is not good.

For now we can leave it as it as is without adding the `TEUCHOS_DEPRECATED` macro and therefore the deprecated warning will not appear.  But we are creating a time bomb that will eventually go off when Trilinos does its next major release and follows the process at:

* https://tribits.org/doc/TribitsDevelopersGuide.html#physically-removing-deprecated-code

to actually remove the deprecated code.  And for any customers that are configuring Trilinos to remove deprecated code, we need to provide them a smooth path to upgrade.  That path, I think, should be to provide `<class_header_name>OStream.hpp` and let them include that.  But we can add that in a future PR when needed.

I will go ahead and approve the PR.

Reply:
With respect to users removing dependence on this deprecated code, I see two options:

- clients supply their own operator<<(ostream&) within their native namespaces, if desired, for the container types they employ.  They could choose to have this trivially invoke obj.print() in a minimalist solution, or implement a prettier print than what's provided (especially if the output is intended for something other than debug dumps/exception handling).  This is what we have done for Dakota.

- The other alternative would be for Teuchos to provide an operator<<(ostream&) in the global namespace, in order to provide a default implementation than can be overridden. Neither you nor Mark seemed keen on this idea, which is fine with me as I personally prefer the previous option.  The only reason to consider this is if we want to reduce the deprecation burden.

Reply:
@mseldre and @rhoope, I think the best approach is to new `<class_header_name>OStream.hpp` files and define ostream operator ifdefed as:

```
#ifdef TEUCHOS_HIDE_DEPRECATED_CODE
namespace Teuchos {
// \brief Print the given SerialBandDenseMatrix to the given output stream.
template<typename OrdinalType, typename ScalarType>
std::ostream& operator<< (std::ostream& os, const Teuchos::SerialBandDenseMatrix<OrdinalType, ScalarType>& obj)
{
  obj.print (os);
  return os;
}
} // namespace Teuchos
#endif // TEUCHOS_HIDE_DEPRECATED_CODE
```

Then customers that are building with deprecated code disable just need to include `<class_header_name>OStream.hpp` and they will compile with or without `TEUCHOS_HIDE_DEPRECATED_CODE` defined.  Then we the deprecated code is removed, they will continue to compile just fine.